keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +92,324 @@ with the unreasonable development overhead of a webhook.
-   update validation in the request pipeline, with these exceptions:
-   - for fields with `x-kubernetes-immutable-keys: true` only the keys of the map or array (specified via `x-kubernetes-list-map-keys`) are compared.
-   - for fields with `x-kubernetes-list-type: set` the order is ignored.
+1. adding two optional vendor extensions to CRD OpenAPI schemas:
+ 
+   1. `x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly` and
+   2. `x-kubernetes-key-mutability: Immutable | AddOnly | RemoveOnly`.
+   
+   The second is restricted to `Immutable` *if* the schema is of type `object` and `properties` is set, the first is restricted otherwise.
+   
+   At the root-level and inside `.metadata` of an object, `x-kubernetes-mutability` and `x-kubernetes-key-mutability` are forbidden.
+   
+   Both extensions can only be set in `v1` CRDs on creation, but updated in `v1beta1` if preexisting.
    
-   If that comparison fails, the request is rejected as 400 "Bad Request".
+2. to use **strict deep-equal** comparison of those fields marked as immutable during
+   update validation in the request pipeline, with the only exception outlined below in the detailed semantics of the different values of the two vendor extensions.
    
 We create another KEP to define custom normalization steps for CRs done during 
 deserialization from etcd and when receiving a request (just after pruning and defaulting).
 
-### OpenAPI extension `x-kubernetes-immutable`
+### Recursive Scope
+
+If the `x-kubernetes-mutability` vendor extension is set (to `Immutable`, `AddOnly`, `RemoveOnly`), `x-kubernetes-mutability: Immutable` is implicitly applied recursively to all subschemas.
+
+The `x-kubernetes-key-mutability` vendor extension does not apply recursively.
+
+### Semantics of `x-kubernetes-mutability`
+
+#### Immutable list, arbitrary list type
+
+The whole list is immutable in the strict sense of `reflect.DeepEqual`.
+
+#### Immutable map / object, arbitrary map type
+
+The whole map is immutable in the strict sense of `reflect.DeepEqual`.
+
+#### Mutable object, immutable fields
+
+The value of a field is immutable, and it cannot be added (Immutable, RemoveOnly) or removed (Immutable, AddOnly).
 
-The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
-arrays and objects. It recursively applies to the whole subtree:
+*Example 1 - Immutable*:
+
+```yaml
+properties:
+  foo:
+    type: string
+    x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly
+```
+
+Allowed for AddOnly, disallowed otherwise:
+- `{}` → `{foo:"a"}`
+
+Allowed for RemoveOnly, disallowed otherwise:
+- `{foo:"a"}` → `{}`
+
+Disallowed:
+- `{foo:"a"}` →	`{foo:"b"}`
+
+#### Mutable array, immutable items, list type list
+
+The items with their respective index in the array are immutable. Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+
+*Example 2*:
 
 ```yaml
 properties:
```

Comment:
what makes this valid now? `x-kubernetes-list-type=list`? `type=array`?

Reply:
fixed. Added all missing types.